File: src/main/kotlin/stellarwitch7/libstellar/registry/cca/CCAComponentRegistrar.kt
package stellarwitch7.libstellar.registry.cca

import org.ladysnake.cca.api.v3.component.Component
import org.ladysnake.cca.api.v3.component.ComponentKey

interface CCAComponentRegistrar<T> : CCAGenericComponentRegistrar {
    val registered: ArrayList<(T) -> Unit>

    fun <C : Component> register(key: ComponentKey<C>, callback: (T, ComponentKey<C>) -> Unit): ComponentKey<C> {
        registered.add { callback(it, key) }
        return key
    }

    fun register(registry: T) {
        for (fn in registered) {
            fn(registry)
        }
    }
}

File: src/main/scala/stellarwitch7/libstellar/registry/cca/CCAComponentRegistrar.scala
package stellarwitch7.libstellar.registry.cca

import org.ladysnake.cca.api.v3.component.Component
import org.ladysnake.cca.api.v3.component.ComponentKey
import scala.collection.mutable.ListBuffer

trait CCAComponentRegistrar[T] extends CCAGenericComponentRegistrar {
  val registered: ListBuffer[(T) => Unit]

  def register[C <: Component](key: ComponentKey[C], callback: (T, ComponentKey[C]) => Unit): ComponentKey[C] =
    registered.append(callback(_, key))
    key

  def register(registry: T): Unit =
    registered.foreach(_(registry))
}
